docs/design/2024-05-10-extension-authentication-plugin.md

```diff
@@ -0,0 +1,237 @@
+
+`VerifyPrivilege`: is called to authorize users for executing SQL statements. This will be an extra check on top of the existing check done with the `MySQLPrivilege.RequestVerification`, so it should be called after the statement here: https://github.com/pingcap/tidb/blob/8d9e67b37dea759db0980aeddf4da967bf93e83e/pkg/privilege/privileges/privileges.go#L189.
+If the user does not have the privileges granted in MySQL, the overall approval will fail and the user will get a privilege error. The overall logic should be:
+```go
+func VerifyPrivilege(mysqlPriv *MySQLPrivilege, user string) bool {
+	return mysqlPriv.RequestVerification(...) && getPluginForUser(user).RequestVerification(...)
+}
+```
+
+`VerifyDynamicPrivilege`: similar to the above interface, but used for dynamic privileges. This will be an extra check on top of the existing check done with the `MySQLPrivilege.RequestDynamicVerification`, so it should be called after the statement here: https://github.com/pingcap/tidb/blob/8d9e67b37dea759db0980aeddf4da967bf93e83e/pkg/privilege/privileges/privileges.go#L129.
+If the user does not have the dynamic privileges granted in MySQL, the overall approval will fail and the user will get a privilege error. The overall logic should be:
+```go
+func VerifyDynamicPrivilege(mysqlPriv *MySQLPrivilege, user string) bool {
+	return mysqlPriv.RequestDynamicVerification(...) && getPluginForUser(user).RequestDynamicVerification(...)
+}
+```
+
+Note that for both the privilege checking functions above, they only receive one privilege in the parameter. This means if the transactions include multiple statements requiring multiple privileges, these functions will be called multiple times using the different privileges.
+
+The above interfaces and parameters should give the users all information they need to make a decision on the authentication and authorization. For authorizations before each request, the MySQL user should already be granted with the corresponding privileges in MySQL using `GRANT` statements.
+
+### Compatibility
+The feature needs to be compatible with existing user related statements and options, such as `CREATE USER`, `ALTER USER`, `SET PASSWORD`.
+However, it will not work with existing functionalities such as password expiry and password validation. Users can implement their own password validation logic inside the `GenerateAuthString` implementation.
+
+It will also not work with `CREATE VIEW SQL SECURITY DEFINER` since the view is created with the privileges of the view creator, and the plugin will not be able to check the privileges of the view creator because it
+does not have the `tls.ConnectionState` information for the definer to pass into `VerifyPrivilege`. This means that if a user is created using an auth plugin, it can only create views using `CREATE VIEW SQL SECURITY INVOKER` since
```

Comment:
I think we can just forbid creating a view by setting `DEFINER` to a plugin user, that is:

- It is NOT allowed for ANY user to create a view by setting `DEFINER` to a "plugin user".
- It is ALLOWED for ANY user to create a view by setting `DEFINER` to a "normal user".

However, when I checked the mysql's behavior and found that the definer will not be verified when creating a view...

```
mysql> create user testuser1;
Query OK, 0 rows affected (0.01 sec)

mysql> create definer=testuser1 view v1 as select * from t;
Query OK, 0 rows affected (0.00 sec)
```

Maybe we can check the `DEFINER` as mysql's behavior to postpone it to select? Anyway, we only need to forbid a plugin user as the definer, that is enough.

Reply:
Moreover, if a plugin does not define `RequestVerification` or `VerifyDynamicPrivilege`, the user that uses this plugin can still be a definer, I think

Reply:
@lcwangchao That makes sense. I have updated the doc to reflect this. Please review again. Thanks!